mm2src/mm2_main/src/lp_swap/swap_watcher.rs

```diff
@@ -0,0 +1,529 @@
+                ]));
+            },
+        };
+
+        let tx_hash = tx.tx_hash();
+        info!("Taker payment spend tx {:02x}", tx_hash);
+        let tx_ident = TransactionIdentifier {
+            tx_hex: tx.tx_hex().into(),
+            tx_hash,
+        };
+
+        let secret = match self
+            .taker_coin
+            .extract_secret(&self.data.secret_hash[..], &tx_ident.tx_hex.0)
+        {
+            Ok(bytes) => H256Json::from(bytes.as_slice()),
+            Err(e) => {
+                return Ok((Some(WatcherCommand::Finish), vec![
+                    WatcherEvent::TakerPaymentWaitForSpendFailed(ERRL!("{}", e).into()),
+                ]));
+            },
+        };
+
+        Ok((Some(WatcherCommand::SpendMakerPayment), vec![
+            WatcherEvent::TakerPaymentSpent(TakerPaymentSpentData {
+                transaction: tx_ident,
+                secret,
+            }),
+        ]))
+    }
+
+    async fn spend_maker_payment(&self) -> Result<(Option<WatcherCommand>, Vec<WatcherEvent>), String> {
+        let spend_fut = self.maker_coin.send_taker_spends_maker_payment_preimage(
+            &self.data.taker_spends_maker_payment_preimage,
+            &self.r().secret.0.clone(),
+        );
+
+        let transaction = match spend_fut.compat().await {
+            Ok(t) => t,
+            Err(err) => {
+                if let Some(tx) = err.get_tx() {
+                    broadcast_p2p_tx_msg(&self.ctx, tx_helper_topic(self.maker_coin.ticker()), &tx, &None);
+                };
+
+                return Ok((Some(WatcherCommand::Finish), vec![
+                    WatcherEvent::MakerPaymentSpendFailed(ERRL!("{}", err.get_plain_text_format()).into()),
+                ]));
+            },
+        };
+
+        broadcast_p2p_tx_msg(
+            &self.ctx,
+            tx_helper_topic(self.maker_coin.ticker()),
+            &transaction,
+            &None,
+        );
+
+        let tx_hash = transaction.tx_hash();
+        info!("Maker payment spend tx {:02x}", tx_hash);
+        let tx_ident = TransactionIdentifier {
+            tx_hex: transaction.tx_hex().into(),
+            tx_hash,
+        };
+
+        Ok((Some(WatcherCommand::Finish), vec![WatcherEvent::MakerPaymentSpent(
+            tx_ident,
+        )]))
+    }
+}
+
+impl AtomicSwap for Watcher {
+    fn locked_amount(&self) -> Vec<LockedAmount> { unimplemented!() }
+
+    #[inline]
+    fn uuid(&self) -> &Uuid { &self.uuid }
+
+    #[inline]
+    fn maker_coin(&self) -> &str { self.maker_coin.ticker() }
+
+    #[inline]
+    fn taker_coin(&self) -> &str { self.taker_coin.ticker() }
+
+    #[inline]
+    fn unique_swap_data(&self) -> Vec<u8> { unimplemented!() }
+}
+
+pub async fn run_watcher(swap: RunWatcherInput, ctx: MmArc) {
+    let swap_ctx = SwapsContext::from_ctx(&ctx).unwrap();
+    let uuid = swap.uuid().to_owned();
+    let mut attempts = 0;
+
+    let swap_lock = loop {
+        match SwapLock::lock(&ctx, uuid, 40.).await {
```

Comment:
No watchers don't save any swap info to a file yet, I'll add that feature in the next sprint. So I removed the relevant parts.

Reply:
I was thinking that if a watcher stops running and starts again, it can continue watching for the maker spending taker payment. The same way as the taker/maker swaps do. We can discuss if this is necessary of course.